### DISCUSSION TITLE
Loss function in GNN

### DISCUSSION THREAD

**Discussion by AlizadehF** [pyg-team/pytorch_geometric#9526] [Q&A] [answered]
Hello,
How can use this loss function for HGNN with GraphSAGE:

![lagrida_latex_editor](https://github.com/user-attachments/assets/5395431a-41b8-4aad-9d12-8d795638d2e3)

Does this loss propagate on each node in each subgraph?

And, how can add edge attribute to HGNN model? I have both node and edge features (Hetero Data).

**Comment by rusty1s** [comment, marked as answer]
You can just implement this loss in vanilla PyTorch on the node embeddings. Gradients are then backpropagated to all nodes that are reachable within the GNN.

If you want to use a GNN layer that supports edge features, take a look at https://pytorch-geometric.readthedocs.io/en/latest/cheatsheet/gnn_cheatsheet.html
